### PR TITLE
Added CountMinSketch

### PR DESCRIPTION
**Pull request by ashtul**

### CODE REVIEW

**Comment by mnunberg** [comment]
I haven't looked at all the code yet, but please place all third party stuff (i.e. stuff that isn't original code of redisbloom) inside the `contrib` directory

File: src/cms.c [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+#include "../rmutil/rmutil_strings.h"
+
+#define CMS_SIGNATURE "COUNTMINSKETCH:1.0:"
+#define CMS_SEED 0
+#define MAGIC 2147483647  // 2^31-1 according to the paper
+#define min(a, b) ((a) < (b) ? (a) : (b))
+
+typedef struct CMSketch {
+    long long counter;
+    int depth;
```
Comment (mnunberg) [inline src/cms.c]: use unsigned (e.g. size_t) for size variables.

File: src/cms.c [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+            (prob < 0) || (prob >= 1)) {
+        RedisModule_ReplyWithError(ctx, "ERR invalid probabilty");
+        return REDISMODULE_ERR;
+        }
+        width = ceil(2 / error);
+        depth = ceil(log10f(prob) / log10f(0.5));
+    }
+
+    CMSketch* cms = NewCMSketch(key, width, depth); /* = NULL?, top or here */
+    if(NULL == cms) {
```
Comment (mnunberg) [inline src/cms.c]: space

File: src/cms.c [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+    if (argc != 4) {
+        return RedisModule_WrongArity(ctx);
+    }
+    RedisModule_AutoMemory(ctx);
+
+    RedisModuleKey *key =
+        RedisModule_OpenKey(ctx, argv[1], REDISMODULE_READ | REDISMODULE_WRITE);
+
+    /* Verify that the key is empty. */
+    if (RedisModule_KeyType(key) != REDISMODULE_KEYTYPE_EMPTY) {
```
Comment (mnunberg) [inline src/cms.c]: Why not define a custom module type?

Comment (ashtul) [reply src/cms.c]: Can you please expand on this?

**Comment by ashtul** [comment]
Hi Mark,

Here is a pull request for the separated CMS.

I will work on TopK and send a separate request for it.

BR

**Pull request closed without merging**